### ISSUE TITLE
Extension (plugin) support

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#30]
There should be support for writing plugins which are more complex than the simple (non-js) userscripts. Python plugins are the way to go.

# FAQ

**What's the current state of this issue?**

~~September 2018: I'm working on this [as a student research project](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2018-September/000051.html). The main focus is designing a clean plugin API which can then be used to move internal qutebrowser parts to "plugins". As soon as this work is done (late 2018), the next step is gradually opening up the plugin API to third-party plugins.~~

~~March 2019: The research project went quite well ([paper](https://github.com/qutebrowser/qutebrowser-hsr-sa/releases/download/final/sa.pdf)), but also was quite academic - not that much time was actually spent on implementation. As a result, there's now an initial extension API used internally in qutebrowser, but it's far from finished and not open to third-party extensions yet.~~

~~I likely won't have much time to continue working on extensions until my bachelor thesis is finished in mid-June (see the [announcement mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-March/000058.html) for details),  but after that I plan to launch another crowdfunding focused on long-term donations (Patreon/OpenCollective). This should allow me to work 2-3 days per week on qutebrowser for a longer time, primarly focused on finishing extensions.~~

~~September 2019: See [this announcement mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-March/000058.html) for a longer recap of how the research project went. I finished my Bachelor thesis and had my final exams in late August. This means my studies are now finished and I finally have more time for qutebrowser on my hands again. Like outlined in that announcement mail, I'm now employed around 16h/week, while I plan to spend the rest of the time on qutebrowser and some freelancing work. At the moment, I'm planning the next crowdfunding campaign (also see [the related survey](https://thecompiler.typeform.com/to/RTE31y)) and getting things set up. After things are rolling, there are some things which are currently more urgent than plugin support: Making sure things continue to work with the latest Qt 5.13/5.14 versions and taking care of [a lot of contributions](https://github.com/qutebrowser/qutebrowser/pulls). Once that backlog is down and some other work I started is finished, I finally plan to get back to extension support again, this time with enough resources to get things to a working state!~~

October 2019: See the recent [roadmap mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-October/000066.html) for more details on what's going on in qutebrowser development - this definitely isn't forgotten, there are just some more important things to look at first. I also launched the [next crowdfunding](https://github.com/sponsors/The-Compiler) which will allow me to work on qutebrowser part-time (without having to write a research paper this time :laughing:) and hopefully accelerate things a lot.

**How can I get updates?**

Subscribe to this issue. You might also want to subscribe to the [qutebrowser-plugins repository](https://github.com/qutebrowser/qutebrowser-plugins) and the [qutebrowser-hsr-sa repository](https://github.com/qutebrowser/qutebrowser-hsr-sa) where the related university documentation will land (also check the [releases](https://github.com/qutebrowser/qutebrowser-hsr-sa/releases) if you're curious).

**What about JavaScript plugins / WebExtensions?**

Support for WebExtensions is currently out-of-scope for qutebrowser. It's hard to even evaluate to which extend support for them would be possible, but if possible, it's going to be a lot of work. Partial support for often used plugin APIs (i.e. *some* WebExtensions) might happen some day, but currently it's too early to say if/when. The main focus is on a Python plugin API right now. Related QtWebEngine issues which might make this easier: [QTBUG-51185](https://bugreports.qt.io/browse/QTBUG-51185), [QTBUG-61676](https://bugreports.qt.io/browse/QTBUG-61676).

Also, a Python plugin API will allow for more freedom, since we treat plugins as trusted code.

**Why don't you just allow people to inject Python code into qutebrowser?**

Firefox had the issue of having a plugin API which tightly integrates with its internals, and had to drop it because it made further Firefox development harder, and bigger changes impossible. Also, every Firefox update broke a lot of plugins. This is not what I want things to end up like with qutebrowser.

In theory, you can do some things via a `config.py` file, like done in the [qutenyan](https://gitlab.com/jgkamat/qutenyan) project. However, this is entirely unsupported and discouraged.

**Do you call them plugins or extensions?**

They used to be called plugins - however, since that term is often used to refer to NPAPI/PPAPI plugins (like Flash) in the context of web browsers, they were later renamed to extensions.

# Plugin ideas
- uMatrix: #28
- More sophisticated adblocking: #29 (host-blacklist based adblocking is implemented in qutebrowser's core)
- HTTPS everywhere: #335
- Ghostery/Disconnect
- Certificate Patrol
- (random) user agent switcher
- Mouse gestures (can be done via `easystroke` to some degree)
- Remove google redirects
- Integration with [LanguageTool](https://www.languagetool.org/)
- Filling forms (e.g. password management with KeePassX/gnome keyring/passwordstore/... - as in #180, also form filling as in #695)
- HoverZoom
- speed dial
- periodic auto-reload
- detailed page loading bar (like otter/opera)
- automatically spawn mpv on video pages
- Find video URLs on a page, like [Media Sniffer](https://addons.mozilla.org/en-US/firefox/addon/media-sniffer/) or [Video DownloadHelper](https://chrome.google.com/webstore/detail/video-downloadhelper/lmjnegcaeklhafolokijcfjliaokphfk)
- copy URLs of all pages or write them to a file
- DownThemAll
- Google Cloud Print ([devguide](https://developers.google.com/cloud-print/docs/appDevGuide))
- Rewriting URLs to their .onion equivalents based on a list
- Chromecast integration? (#1319)
- autoscroll
- [decentraleyes](https://github.com/Synzvato/decentraleyes)
- URL rewriting like PureURL for Firefox (removing things like `utm_foo`)
- Open certain links (like youtube) in a handler application automatically instead of following it
- Adding a "real" address bar
- Adding a bookmarks bar
- Yanking a short URL (e.g. youtube has a shortlink ref for youtu.be links)
- Restarting failed downloads after a timeout (with bad internet connection)
- [alternative stylesheet](https://www.w3.org/Style/Examples/007/alternatives.en.html) selector
- A way to show md5/sha/etc. for downloads or even enter the correct one.
- Plasma Shell browser integration (#3996)
- Exposing the plugin API as an RPC interface usable from other languages
- Integrating [Wallabag](https://wallabag.org/) or other bookmark managers
- Back/Forward by domain instead of page
- Controlling Spotify/Youtube via [MPRIS](https://www.freedesktop.org/wiki/Specifications/mpris-spec/): #5023
- Middle-click scrolling (like on Windows), also see #7215 
- [Noiszy](https://noiszy.com/) (Background tab creating "browsing noise")
- also see existing dwb userscripts
  - [Arch wiki](https://wiki.archlinux.org/index.php/Dwb#Userscripts)
  - [tlvince/dwb-userscripts](https://github.com/tlvince/dwb-userscripts)
  - [dwb-extensions on GoogleCode](https://code.google.com/p/dwb-extensions/downloads/list)
  - [built-in extensions](https://bitbucket.org/portix/dwb/src/5134b576a6ca3282313a20b72ccf5eba4554a2bf/extensions/?at=master)
  - [dwb-extensions repo](https://bitbucket.org/portix/dwb_extensions/src)
- Hooking into notifications
- Showing the current mode in something like i3bar
- Prompting for redirects (?), see #4534
- [TabFS](https://omar.website/tabfs/)
- Mouse navigation similar to keynav, see #6169

# Abilities that plugins should have
- Interact with the `WebElement`s on the page (e.g. make non-clickable links clickable) - or maybe only do this via JS, e.g. [Linkify](http://soapbox.github.io/linkifyjs/).
- Add widgets to the statusbar. (e.g. RSS or search engine symbol)
- Register new commands (with capital letter, like vim)
- Register new settings (in a special `[plugin.foo]` section probably)
- Subscribe to Qt signals (`QWebPage`/`QWebView`)
- Add new completions, e.g. with an external bookmark manager like [cym13/bookmark](https://github.com/cym13/bookmark)
- Create widget-tabs when #724 is done, so someone could theoretically write an RSS reader or mail client or whatever
- Adjust page content, like [Detox](http://usedetox.com/) (and many others)
- Maybe make it possible to interact with the command input so that adding `www.` and `.com` to an URL like in #1749 would be possible.
- Maybe bind keys with prefixes, e.g. `g#` with any number
- Use `acceptNavigationRequest`, e.g. to automatically open mpv if a YouTube link is opened.
- Custom URL schemes
  - IPFS (#1314)
  - Gopher (#321)
  - dat
  - ~~FTP for QtWebKit (#235)~~
- Custom completions (see #3376)
- There might be some more ideas in https://omar.website/tabfs/, as a power user focused interface to the web
- provide custom hinting logic (some ideas in https://github.com/qutebrowser/qutebrowser/issues/674)
- Cloudflare Privacy Pass (#7945)
- Provide custom format strings for commands and tab/window titles, and replace/modify existing format strings (eg a possible solution to https://github.com/qutebrowser/qutebrowser/issues/7685)

# Libraries
- [Yapsy](http://yapsy.sourceforge.net/)
- [straight.plugin](https://github.com/ironfroggy/straight.plugin)
- [stevedore](https://github.com/dreamhost/stevedore)
- [pluggy](https://pypi.python.org/pypi/pluggy)
- [PluginBase](http://pluginbase.pocoo.org/)
- [A simple plugin framework](http://martyalchin.com/2008/jan/10/simple-plugin-framework/)
- [Plug n' play](https://github.com/daltonmatos/plugnplay#readme)
- [anki.hooks](https://github.com/dae/anki/blob/master/anki/hooks.py)
- [pike](https://github.com/pyarmory/pike)

# Resources
- [Fundamental concepts of plugin infrastructures](http://eli.thegreenplace.net/2012/08/07/fundamental-concepts-of-plugin-infrastructures/)
- [The little manual of API design](http://www4.in.tum.de/~blanchet/api-design.pdf)
- [Python API checklist](http://python.apichecklist.com/)
- http://sedimental.org/plugin_systems.html
- [Initial WebExtension support in Epiphany](https://gitlab.gnome.org/GNOME/epiphany/-/merge_requests/866)

# Javascript plugins
- Javascript: [QJSEngine](http://doc.qt.io/qt-5/qjsengine.html)
- [PyV8](https://code.google.com/p/pyv8/) would probably work as well for javascript extensions (and maybe be easier/more pythonic), but it's another dependency and a different javascript engine.

It's probably not possible to directly use QtWebEngine's JavaScript engine, as that's tightly connected to QWebEnginePage actually showing a website.

# Possible API inspirations
- `WebExtensions` API in Chrome/Opera/(soon) Firefox
  - [Mozilla Wiki](https://wiki.mozilla.org/WebExtensions)
  - [Chromium docs](https://developer.chrome.com/extensions)
  - [Opera docs](https://dev.opera.com/extensions/)
- [splinter driver API](https://github.com/cobrateam/splinter/blob/master/splinter/driver/__init__.py)
- [dwb's javascript API](http://portix.bitbucket.org/dwb/api_unstable/)
- Spyder IDE:
  - https://github.com/spyder-ide/spyder/blob/master/spyder/api/plugins.py
  - https://github.com/spyder-ide/spyder/blob/master/spyder/otherplugins.py
  - https://github.com/spyder-ide/spyder/tree/master/spyder/plugins
  - https://github.com/spyder-ide/spyder/wiki/User-plugins
  - https://github.com/spyder-ide/spyder/wiki/Writing-Plugins
  - https://github.com/spyder-ide/spyder-plugin-cookiecutter

## Other interesting links
- [ghacks-user.js](https://github.com/ghacksuserjs/ghacks-user.js) and its [wiki](https://github.com/ghacksuserjs/ghacks-user.js/wiki) for privacy related features.
- [Another user.js](https://github.com/pyllyukko/user.js), dito.

# First ideas
## Separate namespace for plugins

There should be a new `qutebrowser.pluginapi` namespace. Everything outside of that namespace is to be considered to be "private" for plugins, and will likely break at some point.

Only some stuff (like possibly `utils.urlutils`) will be exposed directly - everything else will be exposed through a wrapper, for various reasons:
- I can change stuff without breaking plugins, if I update the wrappers accordingly.
- The wrappers will be easier to expose to a javascript plugin API
- They will be less complex than the qutebrowser internals, so it should be easier to get started

If something is missing from the wrappers, it should be trivial for a plugin author to open a feature request to have it added - this should be encouraged in the docs.
## Qt signal based workflow

I don't think using something like pluggy will be neccessary (especially because pluggy is still expected to undergo changes, and is not packaged).

Instead of that, there's a simple function (say, `plugin_init`) which gets called when a plugin is loaded. The plugin gets a `PluginContext` object with some version information and a `signals` attribute.

The `signals` object then has many signals (perhaps also in sub-objects), like `started`, `cmd.entered`, etc. etc. Such a `signals` object also exists per-tab and per-window. The plugin will get the window-specific object with a global `new_window` signal (and something similar for new tabs).

Those events/signals are the main entry point for plugins. There will also be APIs for certain tasks (like adding an icon to the statusbar), but the majority of tasks should be possible using those signals.

For most _actions_, an API to call commands nicely (i.e. as functions rather than as strings) would probably already suffice.
## Exposing plugin signals to userscripts

This is related to #902 - the signals mentioned above should be written to userscript FIFOs so those can react on the events - similar to [herbstluftwm's](http://www.herbstluftwm.org/) hooks and `herbstclient --idle`.

This could possibly be implemented as a plugin by itself, which binds to all available signals - some layer to transform the signal arguments to strings will be needed either way.

A line written to the FIFO could look like this:

```
category sender signalname arg1 arg2 ...
```
- `category`: For extensibility - currently always `signal`
- `sender`: A dotted path of the sender of the signal, i.e. which `signal` object from above emitted the signal:
  - `global`: A global signal, like `started`
  - `win0`: A signal from the window with win ID 0
  - `win0.tab1`: A signal from the tab 1 in window 0
- `signalname`: A dotted path of the signal, e.g. `started` or `cmd.entered`, etc.
- `arg1`, `arg2`: Arguments which were passed to the signal, which need to be converted to a string in some way. Either there's a plugin which handles this argument for argument, or there's type-checking:
  - `str`: passed through
  - `int`, `float`: `str()` called on them
  - `QUrl`: `url.toString()` called on them
  - window objects -> `win0`
  - tab objects -> `win0.tab1`
# Resources from `BPL`
## resources
- http://www.se-radio.net/2008/07/episode-104-plugin-architectures/
- https://github.com/The-Compiler/qutebrowser/issues/30
- http://www4.in.tum.de/~blanchet/api-design.pdf
- http://eli.thegreenplace.net/2012/08/07/fundamental-concepts-of-plugin-infrastructures/
- http://martyalchin.com/2008/jan/10/simple-plugin-framework/
- https://www.sublimetext.com/docs/3/api_reference.html
- https://ep2015.europython.eu/conference/talks/the-hook-based-plugin-architecture-of-pytest
- https://www.sk89q.com/2010/03/a-first-class-extensible-python-plugin-framework/
- https://trac.edgewall.org/wiki/TracDev
- http://docs.openstack.org/developer/stevedore
- http://aroberge.blogspot.com.es/2008/12/plugins-part-1-application.html
- https://pymotw.com/2/abc/
- http://virtualplants.github.io/latest/dev/peps/pep_plugins.html
## libraries & apps
- https://pypi.python.org/pypi/extensions
- https://www.riverbankcomputing.com/software/dip/download
- https://github.com/kovidgoyal/calibre
- http://yapsy.sourceforge.net/
- http://pluginbase.pocoo.org/
- https://github.com/benhoff/pluginmanager
- https://github.com/daltonmatos/plugnplay#readme
- https://github.com/ccoughlin/NDIToolbox
- https://github.com/dae/anki/
- https://github.com/ironfroggy/straight.plugin
- https://pypi.python.org/pypi/pluggy
- https://github.com/edgewall/trac
- https://github.com/HackEdit
- http://docs.enthought.com/envisage
- https://github.com/openstack/stevedore
- https://picard-docs.musicbrainz.org/en/appendices/plugins_api.html
- https://github.com/PyUtilib/pyutilib/ <----

**Comment by The-Compiler** [comment]
[PyV8](https://code.google.com/p/pyv8/) would probably work as well for javascript extensions (and maybe be easier/more pythonic), but it's another dependency and a different javascript engine.

**Comment by The-Compiler** [comment]
Alternatives for python plugins: [straight.plugin](https://github.com/ironfroggy/straight.plugin), [stevedore](https://github.com/dreamhost/stevedore).

**Comment by The-Compiler** [comment]
[The little manual of API design](http://www4.in.tum.de/~blanchet/api-design.pdf) is a good read too.

**Comment by angelic-sedition** [comment]
> automatically spawn mpv on video pages

Just to contribute to this idea for a possible future plugin, it would be awesome to be able to bind a key to manually invoke mpv (or another player with `:spawn`) on the direct link (not the {url}) of a playing video. With pentadactyl I do this with the firefox media sniffer plugin currently.

**Comment by cwmke** [comment]
Just thought I would throw a couple possibilities/alternatives into the mix.  Ublock Origin, Umatrix, Random-Agent-Switcher.  Thanks for this incredible project.  I'm really excited for it's future.

**Comment by lamarpavel** [comment]
@cwmke The integration of password managers has come up in several suggestions before and it is unlikely that one specific password manager will be integrated into qutebrowser, rather there might be an abstract interface that can be used with many different password managers.

Until then, there are user scripts for this purpose, for example [this one](https://github.com/The-Compiler/qutebrowser/pull/873).

**Comment by The-Compiler** [comment]
I added some thoughts to "First ideas" above - any feedback would be much appreciated!

**Comment by traverseda** [comment]
Has any progress been made in this direction?

**Comment by The-Compiler** [comment]
No, not yet. Tests, the QtWebEngine backend, config revolution and per-domain settings are the current main focus. After that I'll look at this, but it's probably going to take a few months.

**Comment by traverseda** [comment]
Seems like reasonable priorities. Just though I'd drop in and remind you that this issues exists.

A lot of projects, like django, allow you to include apps that are basically just python code. You can get a long way with monkey patching, and just providing some entry points.

Is there a reason you don't provide a quick and hacky solution? Load and run python scripts from a directory or something?

**Comment by The-Compiler** [comment]
Yes, because I don't like quick and hacky solutions :wink:

I don't want third party plugins which break with every release, there's already Firefox for that :wink: - so I'd like to do things right.

**Comment by traverseda** [comment]
Fair enough

**Comment by btall** [comment]
:+1: ``Ghostery/Disconnect`` and ``Remove google redirects``

**Comment by RomanSC** [comment]
+ Default adblock needs have ability to add third party block lists such as uBlocks, otherwise it's not adequite enough to cross out adblock: https://github.com/gorhill/uBlock/tree/master/assets

**Comment by RomanSC** [comment]
+ Default adblock needs have ability to add third party block lists such as uBlocks, otherwise it's not adequite enough to cross out adblock: https://github.com/gorhill/uBlock/tree/master/assets

**Comment by The-Compiler** [comment]
@RomanSC: I think you mean the ability to use adblock-style lists rather than just hosts, because all of the default block lists are already third party, and you can add your own :wink:

See #29 for that, though it's not a priority for me currently, as the simple host-blocker we have now works well enough for 95% of the cases.

**Comment by deniss-s** [comment]
I would add HTTPS Everywhere to the list of useful plugins.

**Comment by haasn** [comment]
Allow me to add FireGestures to the list of my really-want-to-have-this plugins (mostly due to muscle memory, and because I often want to alternate between using the mouse and the keyboard depending on what is more convenient at the time) :)

**Comment by Elronnd** [comment]
[Here](http://lpaste.net/175164)'s what I thought a simple plugin might look like, as inspiration for when/if plugins are actually implemented.  Aside from that, here are some methods, not really organized, that might should be in here:

* `currpage.add_css_rule(tag, ruleslist) # str, dict`

* `currpage.del_css_rule(tag, key, value) # useless, maybe, because we can currpage.add_css_rule(tag, {myrule: default})

* `currpage.move(direction) # (left|right)`

* `currpage.forms[0].fill(content) # str`

* currpage is an instance of Page(), and pages is a list of those?

* `settings.store/fetch(value) # stored in DB? json?  All values that can be stored must be declared, and maxsize declared at beginning of plug, or in manifest.json type thing?  Maybe not manifest.json, but have the user confirm (e.g. "this plug may take up to X MB of space")?

**Comment by The-Compiler** [comment]
So I disagree with some of that, and agree with some other points of that - but we're really still at the brainstorming state here :wink:

Here is how I'd perhaps imagine a plugin reloading a given site periodically:

```python
"""Module docstring automatically used as plugin description."""

from qutebrowser import pluginapi

@pluginapi.hook('new-tab')
def handle_new_tab(tab):
    if tab.url() == QUrl('http://example.com'):
        pluginapi.utils.setup_timer(60 * 1000, tab.refresh)
```

The hooks will then be named just like the same hooks for other purposes (like autocmds).

**Comment by haasn** [comment]
You could borrow a page from the mpv lua API, which I think works quite well for many purposes.

mpv scripts can basically be invoked in three different ways:

1. Through a keybinding or script command they add at load time (which calls a custom function)
2. Through callback from a defined event (similar to the hooks)
3. Triggered by changes to an observed variable

#3 could be merged into #2 in some sense, though, but I still think the ability to explicitly watch named variables (mpv calls them “properties”) can be useful for some purposes. For example, I could imagine a simple history beancounter (assuming we didn't have one already) work by watching for changes to the URL property.

In summary, in addition to thinking what events a plugin could hook, it might also be good to think about what properties a plugin could read/watch.

**Comment by cwmke** [comment]
The WebExtensions route seems like a good move to me since most of the requested extensions already exist for other browsers. This would open up compatibility with a large group of existing extensions. Firefox is moving their extensions system towards this as well.

**Comment by The-Compiler** [comment]
@cwmke I agree WebExtensions support would be awesome, but I don't see it happen anytime soon unless someone finds out how to easily implement it in QtWebEngine. In either case, I definitely want a Python extension API as well.

**Comment by haasn** [comment]
The problem with WebExtensions is that:

1. The API is very limited from what I understand
2. Directly interfacing with qutebrowser elements would be hard from a WebExtensions addon

So in the interest of providing the ability for plugins to, say, define their own commands and do whacky things, a python API would be mandatory (IMO).

But WebExtensions might still be useful for something like μBlock₀ which essentially only interfaces with the underlying browser component - and not so much the python interface.

**Comment by The-Compiler** [comment]
Yeah, I think the main motivation for WebExtension support is that you could run some (all?) existing Chromium plugin - "native" qutebrowser plugins definitely will be in Python.

**Comment by cwmke** [comment]
@The-Compiler is correct in terms of what I am thinking. We would be able to run most (probably not all) Chromium and shortly, Firefox extensions. 

My concern is that by focusing only on Python plugins qutebrowser will be very limited in plugin choices since there will unquestionably be a smaller developer pool to do the creating compared with Chromium and Firefox. This is not a nock on the developers or Python plugins at all since native browser plugins tend to be great. :-)

What creates this concern are similar browser projects like Epiphany, Midori, and Konqueror/Rekonq that have plugin support but only have a few.

Either way I will continue to be a very happy and satisfied qutebrowser user. These are just the remainder of my personal thoughts on the matter.

**Comment by behrmann** [comment]
Having looked around a bit, I agree that some JS-API for plugins would be grand. Unfortunately for the Chrome extension API there seems to be movement only a subset in QtWebEngine for now. 

I would like to see [uBlock](https://github.com/gorhill/uBlock) working in qutebrowser. It looks rather cleanly implemented to me and separates general and platform specific code. I guess it would be a much better solution for adblocking and related wishes than any homegrown solution and it would solve issue #29 in a single stroke.
If there would be a JS-API for plugins the the platform specific code for qutebrowser might even be upstreamed, which would be good PR for qutebrowser, too.

**Comment by Xananax** [comment]
I'm coming late to the party, but I just want to add my +1 to the idea of having js plugins instead of python plugins. I'm not gonna repeat the reasons why, we all know the pros. However I'm not sure I understand the cons properly. Whether python or JS, an interface of limited surface will need to be provided to secure plugins in a sandbox, so it's not like implementing python will be straighforward.  
All I've read is that it is "harder" and might not be "cleaner", and that makes it hard to comment constructively, as I am unable to assess how hard vs the benefits. Can anyone point out some reading material so I can document myself?

From where I stand, it seems to me that if implementing Js plugins is possible at all, then that's how it should be done, as Qute will become instantly much more useful for a lot of people.

**Comment by traverseda** [comment]
I would *much* prefer to write plugins in python. I'm not terribly interested in this project beyond vimperator+firefox outside of that requirement.

Of course greasemonkey style userscripts are valuable too.

**Comment by The-Compiler** [comment]
After talking to @Xananax in IRC, the main misunderstanding was that he thought I wanted to sandbox/restrict plugins in some way, which is indeed difficult/impossible with Python.

That's not the goal however - I view plugin code as trusted code with full privileges, either vetted by qutebrowser devs or by users.

I definitely want a Python API too (it was one of many reasons to write a new thing instead of contributing to an existing vim-like browser!). Nowadays I think a JS API only makes sense for some degree of WebExtensions compatibility (Chrome extensions), and that's definitely more in the future than a Python API is.

**Comment by triclops200** [comment]
I would also like to indicate support of python plugins.

**Comment by s1lvester** [comment]
Okay, so I'm going off on a little rant here: Clearly this project has a) huge potential and b) a lot of interest form the community (hence the number of Issues, PRs, Stars and so on). We need a Plugin API so the community can scratch their individual itches and the core devs can focus on big picture stuff.
@The-Compiler Please make this a priority and give us an API.

**Comment by The-Compiler** [comment]
@s1lvester I don't disagree about needing this, but qutebrowser is a lot of work, so it'll take some time until I get to bigger issues which are quite a bit of work to get right, like this one. It's definitely on my radar (and hey, even I'd like it sooner rather than later!), but currently the last few QtWebEngine issues and the new config system (with per-domain settings) are still more important. Are you aware of [userscripts](https://github.com/qutebrowser/qutebrowser/blob/master/doc/userscripts.asciidoc)?

**Comment by Elronnd** [comment]
I'm pretty sure he wants to finish perfecting webengine first, so that plugins will be interfacing with the final product, so the API can be more stable.

**Comment by tidux** [comment]
There's a JavaScript IPFS implementation that's working towards full IPFS interoperability as a WebExtension, so consider this a vote towards that API once Qutebrowser is ready to begin work on it.

**Comment by The-Compiler** [comment]
I just launched qutebrowser's second crowdfunding! The main goal there is the new config system (#499, #27 and others), but after that we'll be a big step closer to having a plugin API (because all other big things are finally out of the way).

If there's leftover time after the new config is implemented, I'll spend that on finishing QtWebEngine support, and then (time permitting) on some initial work for a plugin API.

Like last year, qutebrowser t-shirts and stickers also are available there.

See the [Kickstarter campaign](https://www.kickstarter.com/projects/the-compiler/qutebrowser-v10-with-per-domain-settings?ref=3kvxwp) for more infos! :tada: :sparkles:

**Comment by Elronnd** [comment]
A lot of websites have buttons that don't have hints.  Another thing addons should be able top do is add hints to arbitrary elements, to "fix" those websites, so to speak.  Or, alternatively, to remove hints from arbitrary elements, in case of a duplicate.

**Comment by elshize** [comment]
Just adding another plugin idea: integration with LanguageTool: https://www.languagetool.org/

**Comment by dvs1** [comment]
Maybe Spyder's system will give you some ideas (it's a PyQt program).

https://github.com/spyder-ide/spyder/blob/master/spyder/api/plugins.py
https://github.com/spyder-ide/spyder/blob/master/spyder/otherplugins.py
https://github.com/spyder-ide/spyder/tree/master/spyder/plugins
https://github.com/spyder-ide/spyder/wiki/User-plugins
https://github.com/spyder-ide/spyder/wiki/Writing-Plugins

It looks like they are trying to use Cookiecutter "for jumpstarting Spyder IDE plugins quickly":
https://github.com/spyder-ide/spyder-plugin-cookiecutter

🖖

**Comment by Anachron** [comment]
@The-Compiler now that `1.0.(0|1)` is out I hope we can start with this feature (after the domain depending config). 

I would volunteer to try to get uBlock, Youtube Center and alike going.

**Comment by The-Compiler** [comment]
@Anachron Big things like this most likely won't happen out of my holidays. I wouldn't hope for plugin API support before ~July 2018. And as usual, I don't really do ETAs for hobby work, so take this with a grain of salt.

**Comment by The-Compiler** [comment]
Oh, and this:

> I would volunteer to try to get uBlock, Youtube Center and alike going.

sounds like you're referring to WebExtension support. That's even further off into the future, if possible at all.

**Comment by jgkamat** [comment]
> ... uBlock, Youtube Center and alike going.

FWIW, qutebrowser has a built-in adblocker (which is customizable), has functionality for `:spawn`ing mpv to open youtube videos (which is a much better experience, IMHO), and a [greasemonkey PR](https://github.com/qutebrowser/qutebrowser/pull/3040) is open which might let you do what you want.

**Comment by Anachron** [comment]
I already use mpv for videos and the adblocker is a hostblocker which has issues because some pages dont work at all with it.

Oh and I cant jump to a lot of google results because the hosts are blocked and I cant add a script to remove the proxy (tracking) url.

**Comment by Elronnd** [comment]
Greasemonkey PR is certainly interesting (and if it can connect to local websockets, opens up for some interesting integration with userscripts).  But I wonder if it could be rewritten as a plugin once plugins are a thing?  As for ublock, it might make more sense to make a plugin that takes ublock filter lists as opposed to just host-based lists (which could also be implemented in core -- I suggested this a while back, got told it was probably possible, and then nothing happened?)

**Comment by lamarpavel** [comment]
>edit: That is, unless a contributor does it and takes care of getting it into qutebrowser. It's definitely not something I'd recommend as a first contribution, though.

It doesn't have to be one contributor. The first things that should be done when trying to create a plugin API is to look at other browsers and the most popular plugins there, get an idea of what they would require and write a few lists with details on requirements. All of this should happen before thinking about the architecture of the API and could be done without any Python experience. This kind of research could very well be done by people from the community who already know what plugins they would like to see. Not until after this has been documented clearly should the planning of the API begin. Either someone with a lot of experience and time could then take over, or we sketch out more details in a discussion (where at least some should have deeper knowledge of Python and the qutebrowser code base).

Not sure if a github issue is the right place for this, though.

**Comment by Elronnd** [comment]
I think a github issue would be fine.  Not *this* one necessarily, but in general I think it's be fine.  Some ideas:

* Inject html element/css rule/js into a page

* Set up a script (or function maybe) to run at certain times (every n seconds, on page load (and to be given the page that was loaded), on startup, perpetually in the background)

* Inject commands (like scroll down, close tab, move to the 3rd tab, etc.)

* Add/remap commands

* Control the window (resize, move, minimize)

* Insert a Qt widget (maybe this isn't what they're called, I haven't used Qt) in a new tab.  Or a new window (system window, not qute window)

* Allow for python-js interop (maybe injected scripts can send/receive websocket requests from localhost).  This is especially interesting because it might allow us to implement WebExtensions as a plugin

* Create new network protocols (ftp(s), gopher, what-have-you)

**Comment by alem0lars** [comment]
Is there any way to currently use LastPass in qutebrowser?

**Comment by The-Compiler** [comment]
@alem0lars I have no idea. You could probably copy-paste things, or adjust the [password_fill script](https://github.com/qutebrowser/qutebrowser/blob/master/misc/userscripts/password_fill) to work with LastPass if there's a commandline client or Python library (if you do, please open a pull request!).

**Comment by alem0lars** [comment]
I have interest in contributing to provide LastPass support. I have Python knowledge, however I don't know qutebrowser from the developer's perspective.

Atm is there a way to integrate lastpass in qutebrowser with a plugin or something else (I can write the needed code)?

The passwordfill is just a userscript, that would be the last chance if nothing else is possible.

**Comment by The-Compiler** [comment]
Why "just a userscript"? There currently isn't any plugin API other than that, that's why this issue is still open :wink:

**Comment by laugh-at-me** [comment]
[HTTPS Everywhere](https://github.com/EFForg/https-everywhere) cannot be imitated by a userscript and is essential.

**Comment by laugh-at-me** [comment]
[HTTPS Everywhere](https://www.eff.org/https-everywhere). This cannot be imitated by a userscript and is essential.

**Comment by The-Compiler** [comment]
Getting back to what @lamarpavel and @Elronnd mentioned earlier:

I think it's a good idea to take a look at e.g. the WebExtension API to get some inspiration, but at the same time, there are good reasons for our API to diverge from that:

- It's Python, and not JavaScript. We should create an API which seems right/clean/simple/intuitive to Python programmers rather than try to mirror JavaScript APIs.
- We regard all extensions as trusted, which is wildly different from WebExtensions (where extensions are untrusted and very limited).
- Extensions have Python's and Qt's standard library available, so there's probably things extensions can just take from there.
- We have a lot of concepts in qutebrowser which should be in the extension API, but don't exist that way in WebExtension APIs, and vice versa. For starters, the whole concept of commands and keybindings.

Here's roughly how I'd like to go about this issue:

- Get extension infrastructure itself to work, at first only with qutebrowser-internal "extension" but no public interface.
- Expose the bare minimum of qutebrowser which is feasibly well-designed and stable. Currently, this is probably (part of) the [tab API](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/browser/browsertab.py) and [cmdutils](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/commands/cmdutils.py).
- Move everything possible in qutebrowser's code base to use the newly exposed extension API (I'd probably call this "components" or so, and move all code only using the stable extension API to a new directory in the source tree). This is probably a good time to take care of #3319 as well.
- Check the remaining code in qutebrowser's core, and decide what should be moved out from the core into a component. Generally, everything which can be mapped to a general-purpose stable API (say, the whole adblocking stuff) should probably move to a component, everything which can't (say, the concept of tabs) should stay in the core.
- Gradually add more stuff to the extension API, and move stuff out of the core into "components".

Then at some point, we should already have a nice API, with real usage (qutebrowser itself, basically). Together with the use cases in the original post of this issue, this should form the extension API. As soon as it feels stable enough, it should be officially exposed as an extension API, and then gradually get more power based on what plugins people want to write (and what's reasonable).

What I'd like to avoid is the situation Firefox had with its XUL/XPCOM API - having a very powerful plugin API at the expense of painting ourselves into a corner with what changes we can make to qutebrowser's codebase.

**Comment by NoctuaNivalis** [comment]
I like the roadmap.

> What I'd like to avoid is the situation Firefox had with its XUL/XPCOM API - having a very powerful plugin API at the expense of painting ourselves into a corner with what changes we can make to qutebrowser's codebase.

Main difference: qutebrowser users would understand the need to update their installed/written plugins to a newer API, so we don't need really strict backwards compatability. It's usually OK to break things a bit.

**Comment by lamarpavel** [comment]
I very much like the roadmap you fleshed out, moving stuff out of qutebrowser and into extensions as soon as the core extension API is built is a win/win/win. First, we get to test the API with real use cases, secondly the API will serve as a separation layer to restrict code entanglement (or what OO people sometimes call _cross cutting concerns_) and thirdly, we get an extension API (duh).

For inspiration I can recommend looking at the IRC client weechat, which has internalised this concept to the point that even the IRC related code is a module (that can even be turned off).

And one more thing: I've been somewhat invested in the Firefox 57 debacle and was forced out of my comfort zone due to __all__ of my plugins no longer working (except for httpseverywhere) and then getting stuck with ABI incompatibility of my outdated Firefox 55 and the current icu package from repos (and then the entire dependency hell following that). So I looked a bit around and read through some discussions about the WebExtensions API. Here are some very relevant points of reference that we can learn from:

- Section __End of Life and Alternatives__ of the Vimperator Readme: https://github.com/vimperator/vimperator-labs
- issues of tridactyl regarding webextension limits: https://github.com/cmcaine/tridactyl/issues?q=is%3Aissue+is%3Aopen+label%3Awebext-limitation
- This entire discussion: https://github.com/vimperator/vimperator-labs/issues/705

**Comment by The-Compiler** [comment]
> For inspiration I can recommend looking at the IRC client weechat, which has internalised this concept to the point that even the IRC related code is a module (that can even be turned off).

I planned to take a look at weechat indeed. My main inspirations so far were:

- [pytest](https://docs.pytest.org/en/latest/), which has a [simple yet powerful plugin API](https://docs.pytest.org/en/latest/writing_plugins.html) based on hooks, and default implementations of those hooks shipped with the core.
- Strangely enough, [odoo](https://www.odoo.com/) (an ERP/CRM), which also has quite a small core and [lots of addons shipping with it](https://github.com/odoo/odoo/tree/master/addons).

I just think it's important to not make *too much* pluggable outright. With pytest, pretty much everything is a plugin hook, and plugins can expose their own plugin hooks - but that also gets complicated fast, and makes it harder to refactor stuff.

---

As for the discussions about Vimperator, I've been a part of them since pretty much the beginning :wink: It's also my main source to keep the [lists of alternatives](https://github.com/qutebrowser/qutebrowser#similar-projects) up to date.

**Comment by lamarpavel** [comment]
Oh and one other quick idea I had: A plugin could be added that exposes an RPC interface through something language agnostic and simply translates that to the extension API in Python. So you would expose the API to any kind of external program without having to change something in the API.

**Comment by robehickman** [comment]
Why not implement webextensions to gain access to proprietary addons like last pass? Lack of support for that rules out using any of these alternative browser projects for me.

**Comment by traverseda** [comment]
Webextentions is pretty new, and wasn't an option when work on this feature started. It's probably worth opening a new issue just to discuss them.

**Comment by The-Compiler** [comment]
Please read the thread, or at least search for webextension in this thread :wink:

**Comment by traverseda** [comment]
It might be worth opening a ticket just for web-extention support. Since python-ey plugins and webextentions are very different things with very different.

**Comment by Elronnd** [comment]
Guys.  Webextensions aren't happening.  Give it a rest.

**Comment by traverseda** [comment]
You know what would solve people coming into this thread to talk about webextentions? Opening a ticket with the words "`[feature request]` implement webextentions" and a prominent comment saying "this is a long way off in the future".

Alternatively, changing this tickets title to 'python-based plugin support'. People are going to talk about webextentions, and right now this is the designated place to do that. Designate another place, or make it obvious that this is the designated place to talk about python plugin support.

I'd open such a ticket myself but someone just closed a similar ticket at #3186.

Don't get annoyed when people talk about the most prominent plugin api in the thread about implementing a plugin api, it's going to keep happening. You can control where it happens though.

**Comment by The-Compiler** [comment]
That someone would be me :wink:

Feel free to *discuss* about WebExtensions here, if there's anything new to say. But "WebExtension support would be cool" has been said before, and doesn't change anything about it being far away and maybe not possible at all.

With ~500 open tickets, it makes no sense to have tickets around which aren't actionable at all, which is why there's no dedicated WebExtension ticket, but it fits well in here.

I added a section about WebExtensions to the first post, though.

**Comment by oblitum** [comment]
Just fyi, lastpass does have a terminal client, lpass, which is quite good. I still use Firefox, and use lastpass both as a  Firefox addon as well as lpass in the terminal. It's the sole thing holding me back to switch to qutebrowser, if it the future it manages to use the terminal client, it will be even better than duplicate lastpass functionality via browser extension.

**Comment by pkillnine** [comment]
@oblitum I think support for `lpass` could easily be added to userscript `password_fill` or `qute-pass`.

**Comment by biinari** [comment]
@oblitum I've made a backend to `password_fill` for `lpass`: https://gist.github.com/biinari/e5611220f90258552cd85032c9c09021 - it could be improved upon but works reasonably well for starters.

**Comment by oblitum** [comment]
@biinari cool thanks for that.

**Comment by The-Compiler** [comment]
# Some more notes on my opinion on plugin security

Sandboxing Python in some safe way is widely [regarded as impossible](https://lwn.net/Articles/574215/) - instead, I think it's fine for qutebrowser to trust plugins.

## Manually installed

If people dump stuff in their ~/.local/share/qutebrowser/plugins or whatever it'll be, it's their responsibility to ensure it's safe.

## Plugin manager

For plugins installed via some kind of plugin manager (or something like a `:plugin-install`), there could be a plugin registry somewhere - for example in some JSON/YAML file in the repository, which gets downloaded by qutebrowser with a `:plugin-update` or so. That file could contain entries like:

```
exampleplugin:
  repo: https://github.com/example/qutebrowser-exampleplugin
  version: 1.0
  commit: 11a94957dc038fc27c5ff976197ad2b2d0352d20
```

Then when people write a new plugin or release a new version of an existing one, this happens:

- They push their plugin to a GitHub repository
- They open a PR against qutebrowser to update the plugin index
- A qutebrowser maintainer does a quick review of the plugin code to make sure it doesn't do anything clearly stupid or malicious and merges the PR
- qutebrowser uses that file to offer plugins to install, and when downloading a plugin, makes sure the commit (or possibly some hash on a GitHub `.zip` file, depending on implementation) matches.

That means there's at least some review of what plugins want to do (but people can still install plugins manually). It still should be documented that qutebrowser isn't responsible for whatever happens in plugins.

Feedback welcome! :smile:

**Comment by gebulmer** [comment]
I don't know if it's even worth making a plugin manager, just having a standard directory to dump them is enough. A plugin manager would then be simply another plugin, registering its commands in the usual (provided) way, that handles the others.

**Comment by pkillnine** [comment]
I do like the idea of an official qutebrowser plugin repo that is maintained, since plugins are going to be assumed to be fully trusted. Maybe a simple compromise could be to ship qutebrowser with an official plugin manager, which is simply another plugin (like @gebulmer said), with the official qutebrowser plugin repo? This way if someone wants to use a different plugin manager they can easily remove the default one.

**Comment by traverseda** [comment]
It seems to me like the obvious plugin manager is... pip. It does everything you'd need. I don't know how easy it would be to have a per-user/per-profile virtualenv, but that would certainly make things easier.

**Comment by The-Compiler** [comment]
Yeah, I've had that in the back of my mind for a while as well. I'm not sure how well it'd work with locally installed plugins and with frozen distributions (macOS/Windows), but it's certainly worth some more thought.

**Comment by lamarpavel** [comment]
That would mean pip becomes a qb dependency for regular users?
Also, I don't see how pip makes any guarantees about which plugins were curated by qutebrowser developers.

**Comment by The-Compiler** [comment]
> That would mean pip becomes a qb dependency for regular users?

Indeed, that probably wouldn't be a problem, though. I'm pretty sure it's available on any Linux distribution, and it can be shipped with qutebrowser binaries for macOS/Windows.

> Also, I don't see how pip makes any guarantees about which plugins were curated by qutebrowser developers.

It doesn't - but qutebrowser can do so. If the user invokes pip by hand, it's the same situation like them throwing a plugin somewhere by hand - it's their fault™ :wink:

**Comment by def14nt** [comment]
@The-Compiler This isn't really my field, but wouldn't it be better to use the JavaScript engine that qutebrowser uses, instead of using a different one?

**Comment by The-Compiler** [comment]
@def14nt I'll need more context for that question. The current focus are Python plugins anyways, not JS plugins. As for JS plugins, qutebrowser can't run any JS in QtWebEngine which has more permissions than the page has.

**Comment by def14nt** [comment]
**PLEASE** move on this! These ads are too much, I'm using Chromium until this is fixed, which ruins my i3 workflow.  Really miss tasks, too :cry:

**Comment by The-Compiler** [comment]
@def14nt I [am currently working](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2018-September/000051.html) on this. Comments like this do nothing except making working on qutebrowser more frustrating. You're not in a position to tell me what to do in my spare time.

**Comment by kepi** [comment]
@def14nt If you like something implemented use thumbsup on issue or submit PR with your proposal. But please do not demand something like you did. Its ok to use different browser if QB currently lacks some future you need, you'll have to wait. Btw. if you have problem with ads, try `adblock-update` command, it is good enough currently.

**Comment by def14nt** [comment]
How was I demanding? I said "please". Think you're a *bit* defensive. To be *completely* honest, I was hoping for an update :)

**Comment by The-Compiler** [comment]
I just posted a [roadmap mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-March/000058.html) to the announcement mailinglist - here's the gist of it related to extensions:

The research project went quite well ([paper](https://github.com/qutebrowser/qutebrowser-hsr-sa/releases/download/final/sa.pdf)), but also was quite academic - not that much time was actually spent on implementation. As a result, there's now an initial extension API used internally in qutebrowser, but it's far from finished and not open to third-party extensions yet.

I likely won't have much time to continue working on extensions until my bachelor thesis is finished in mid-June (see the [announcement mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-March/000058.html) for details), but after that I plan to launch another crowdfunding focused on long-term donations (Patreon/OpenCollective). This should allow me to work 2-3 days per week on qutebrowser for a longer time, primarly focused on finishing extensions.

**Comment by The-Compiler** [comment]
I just added an update: See [this announcement mail](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2019-March/000058.html) for a longer recap of how the research project went. I finished my Bachelor thesis and had my final exams in late August. This means my studies are now finished and I finally have more time for qutebrowser on my hands again. Like outlined in that announcement mail, I'm now employed around 16h/week, while I plan to spend the rest of the time on qutebrowser and some freelancing work. At the moment, I'm planning the next crowdfunding campaign (also see [the related survey](https://thecompiler.typeform.com/to/RTE31y)) and getting things set up. After things are rolling, there are some things which are currently more urgent than plugin support: Making sure things continue to work with the latest Qt 5.13/5.14 versions and taking care of [a lot of contributions](https://github.com/qutebrowser/qutebrowser/pulls). Once that backlog is down and some other work I started is finished, I finally plan to get back to extension support again, this time with enough resources to get things to a working state!

**Comment by The-Compiler** [comment]
I originally planned to look at #4665 by @toofar for v2.0.0 (a basic extension loader), but there are still some things in there I want to change (or think about in more detail).

To not delay v2.0.0 any further (and [hopefully get it into the next Debian stable](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=980847)) I'll remove this from the v2.0.0 milestone, but it's definitely still something I want to pick up again as soon as I can.

**Comment by The-Compiler** [comment]
For what it's worth, there is [active work](https://codereview.qt-project.org/c/qt/qtwebengine/+/601275) going on for [WebExtension support](https://bugreports.qt.io/browse/QTBUG-61676) in QtWebEngine.

Feature freeze for [Qt 6.10](https://wiki.qt.io/Qt_6.10_Release) is on May 30th (release planned for September 23rd). That means 9 days to go, let's see if that gets in in time, or if it will have to wait until Qt 6.11 maybe.
